python/cuml/dask/common/base.py

```diff
@@ -18,27 +18,78 @@
 import numpy as np
 from toolz import first
 
-import cudf.comm.serialize  # noqa: F401
+from cuml.dask.common.utils import get_client
+from cuml.dask.common.input_utils import get_datatype
+import pickle
 
 from cuml import Base
 from cuml.common.array import CumlArray
 
 from dask_cudf.core import DataFrame as dcDataFrame
 
-from dask.distributed import default_client
 from functools import wraps
 
 
 class BaseEstimator(object):
 
-    def __init__(self, client=None, verbose=False, **kwargs):
+    def __init__(self, client=None, verbose=False, model=None, **kwargs):
         """
         Constructor for distributed estimators
         """
-        self.client = default_client() if client is None else client
+        self.client = get_client(client)
         self.verbose = verbose
         self.kwargs = kwargs
 
+        self.set_model(model)
+
+    @classmethod
+    def load(cls, file, client=None, verbose=False, pickle_args={}, **kwargs):
+        """
+        Convenience function to load a saved cuml model into a distributed
+        model.
+        """
+        model = pickle.load(file, **pickle_args)
+
+        if isinstance(model, Base):
+            # If serialized model is single GPU, create new
+            # dist model
+            model = cls.__init__(client=client,
+                                 verbose=verbose,
+                                 model=model,
+                                 **kwargs)
+        else:
+            model.client = get_client(client)
+            model.kwargs = kwargs
+            model.verbose = verbose
+        return model
+
+    def __getstate__(self):
+        inst_attrs = self.__dict__
+        inst_attrs["local_model"] = self.get_model()
+        del inst_attrs["client"]
+        return inst_attrs
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+    def get_model(self):
+        """
+        Return trained single-GPU model
+        """
+        local_model = self.local_model
+        if not isinstance(self.local_model, Base):
+            local_model = self.local_model.result()
```

Comment:
The original assumption here was that `self.local_model` was always going to be local. During the TPCxBB process, I found a few issues here. The biggest issue is that I no longer like the assumption that `local_model` (`internal_model`?) is always local because this breaks really badly when a memory pool is used which is occupying most of the memory on the device and the client tries to use a bunch of device memory. I managed to get away from this in the TPCxBB flows by keeping the `internal_models` futures. I think in general we can assume the internal_model is just about always going to be a future.